Add support for playwright  global exceptions to console reporter

In1th opened the pull request:
according to #2244

jan-molak commented:
Hey @In1th, I had a look at how `onError` behaves, and I'd suggest a slightly different approach.

When Playwright Test encounters an unhandled error, it invokes `onError`, and then `onEnd` as usual.
What I'd suggest is to enhance `SerenityReporterForPlaywrightTest` to introduce an optional field `error`:
```ts
export class SerenityReporterForPlaywrightTest implements Reporter {

    private error?: Error;
```
and a simple `onError` handler that assigns any uncaught error to that field:
```ts
    onError(error: Error): void {
        this.error = error;
    }
```
With this capture mechanism, we could then modify `onEnd` like this:
```ts
    async onEnd(): Promise<void> {
        this.serenity.announce(new TestRunFinishes(this.serenity.currentTime()));

        try {
            await this.serenity.waitForNextCue();

            const outcome = this.error
                ? new ExecutionFailedWithError(this.errorParser.errorFrom(this.error))
                : new ExecutionSuccessful()

            this.serenity.announce(new TestRunFinished(outcome, this.serenity.currentTime()));
        }
        catch (error) {
            this.serenity.announce(new TestRunFinished(new ExecutionFailedWithError(error), this.serenity.currentTime()));
            throw error;
        }
    }
```
Since the uncaught error is now correctly emitted as part of the `TestRunFinished` event, `ConsoleReporter` reports it as follows:
```
[test:acceptance] ================================================================================
[test:acceptance] Execution Summary
[test:acceptance] 
[test:acceptance] 
[test:acceptance] Total time: 0ms
[test:acceptance] Real time: 41ms
[test:acceptance] Scenarios:  0
[test:acceptance] ================================================================================
[test:acceptance] ReferenceError: asd is not defined
[test:acceptance]     at /Users/jan/Projects/serenity-js/serenity-js/examples/playwright-test-todomvc/spec/bulk_operations.spec.ts:11:5
[test:acceptance]     at Object.<anonymous> (/Users/jan/Projects/serenity-js/serenity-js/examples/playwright-test-todomvc/spec/bulk_operations.spec.ts:10:9)
```
This is of course better than not having any information about the unhandled error, but not as nice as what Playwright Test `line` reporter outputs:
```
[test:acceptance] ReferenceError: asd is not defined
[test:acceptance] 
[test:acceptance]    at bulk_operations.spec.ts:11
[test:acceptance] 
[test:acceptance]    9 |
[test:acceptance]   10 | describe('Bulk operations', () => {
[test:acceptance] > 11 |     asd
[test:acceptance]      |     ^
[test:acceptance]   12 |     describe('Todo List App', () => {
[test:acceptance]   13 |
[test:acceptance]   14 |         beforeEach(async ({ actor }) => {
[test:acceptance] 
[test:acceptance]     at /Users/jan/Projects/serenity-js/serenity-js/examples/playwright-test-todomvc/spec/bulk_operations.spec.ts:11:5
[test:acceptance]     at Object.<anonymous> (/Users/jan/Projects/serenity-js/serenity-js/examples/playwright-test-todomvc/spec/bulk_operations.spec.ts:10:9)
```

So if you'd be up for that, perhaps "phase 1" of this PR could be to capture and emit the unhandled error, and "phase 2" could be to offer more in-depth feedback to help developers identify what caused the error in the first place?

In1th commented:
@jan-molak, definitely! I'll go and change this to your proposal and try to run tests locally. I suppose I may have an idea of how to get more info, but let me try to run it once 😄

In1th commented:
Ok, I did the version you suggested @jan-molak. I even figured out how to include a code snippet, turns out playwright passes it through its own object.
```
-Error: Trigger failure for worker 0
      -   7 |         beforeAll(async ({ browserName }, workerInfo) => {
      -   8 |             if (workerInfo.workerIndex < 2) {
      ->  9 |                 throw new Error(`Trigger failure for worker ${ workerInfo.workerIndex }`);
      -     |                       ^
      -  10 |             }
      -  11 |         });
      -  12 |
```
I copied it from integration test, since I modified playwright error parser to include code snippet... What means every error has this. That's why I have to ask you the approach, since just adding this functionality on by default would mean changing all failing tests to have this snippet and that sounds brittle.

So I thought about having a reporter config option to enable code snippets in errors.

In1th commented:
@jan-molak I didn't know if I got it right, you want me to implement our own code snippet functionality and go to every console reporter implementation for different frameworks or just do it in the playwright reporter?

For now I added a new flag called `includeSnippetOnError` to include them. Also I added new functionality to playwright tests to be able to manually change ENV variables, for this case a variable to enable snippets just for one test. 😄

jan-molak commented:
>  you want me to implement our own code snippet functionality and go to every console reporter implementation for different frameworks or just do it in the playwright reporter?

ConsoleReporter is generic, so there's only one implementation that works with all the reporters.

To keep it simple, though, I'd suggest splitting this PR into 2:
- one to add the global error handler, as per our conversation
- another to do the formatting

Adding the information about a global unhandled error is already valuable, even if it doesn't have the snippet yet.

WHat do you think?

In1th commented:
Alright, I did the revert to playwright error handling, so I guess for this PR it finally has what it advertises 😆

jan-molak commented on packages/playwright-test/src/reporter/SerenityReporterForPlaywrightTest.ts:
```diff
@@ -40,26 +60,25 @@ import { SERENITY_JS_DOMAIN_EVENTS_ATTACHMENT_CONTENT_TYPE } from './PlaywrightA
-     * - {@apilink SerenityOptions}
-     * - {@apilink SerenityConfig.crew}
-     */
+   * A list of {@apilink StageCrewMemberBuilder|StageCrewMemberBuilders} or {@apilink StageCrewMember|StageCrewMembers}
```
Looks like formatting has gone a bit off; maybe some editor settings that kicked in as part of the commit?

In1th replied:
seems like it, maybe I run auto formatter. Fixed docs are pushed

jan-molak commented:
Merged, will be released in 3.19.0 🚀

Thanks!
